Add packet-io support

Yi-Tseng opened the pull request:
This pull request adds packet-io support.
Also, modify the way we handle stream response messages to support other types of messages.

`usage/packet_io.md` include the way to use the packet-io


fixes #26

Commit:
Add packet-io support

Yi-Tseng commented on p4runtime_sh/p4runtime.py:
```diff
@@ -163,13 +169,23 @@ def stream_recv_wrapper(stream):
+                        self.stream_in_q["arbitration"].put(p)
+                    elif p.HasField("packet"):
+                        self.stream_in_q["packet"].put(p)
+                    elif p.HasField("digest"):
+                        self.stream_in_q["digest"].put(p)
```
I'll add support for digest in the next PR, so I just leave this here

Commit:
Fix python style

antoninbas reviewed (commented):
Thanks for working on this. There's no unit test?

I think this is a good first step, and I have no issue merging this. In the future several improvements are possible IMO: ability to disable packet-ins to avoid growing queue size, ability to filter packet-ins, ability to apply a lambda to packet-ins, etc.

pudelkoM commented on p4runtime_sh/shell.py:
```diff
@@ -2263,6 +2264,106 @@ def add(self, egress_port=None, instance=0):
+        self._md.clear()
+
+
+class PacketIO(_P4EntityBase):
```
Could this class be split up into `in` and `out`?

Seems weird to me to be able to call `send` on a `PacketIn`.

pudelkoM commented on p4runtime_sh/p4runtime.py:
```diff
@@ -198,17 +214,18 @@ def handshake(self):
         start = time.time()
         try:
             while True:
                 remaining = timeout - (time.time() - start)
                 if remaining < 0:
                     break
-                msg = self.stream_in_q.get(timeout=remaining)
+                msg = self.stream_in_q[type_].get(timeout=remaining)
                 if msg is None:
                     return None
```
The loop can be removed as we sort stream messages by type beforehand.

pudelkoM commented:
Thanks for the review @antoninbas!

> ability to apply a lambda to packet-ins

Do you envision a global callback (then invoked in the background) or something more ad-hoc like with `table_entry.read(lambda e: print(e))`?

>ability to disable packet-ins to avoid growing queue size

I guess a global per-queue lambda would also solve the growing queue size problem. The default would just be a discarding function.

>ability to filter packet-ins

Filtering on metadata fields?

antoninbas commented:
@pudelkoM maybe a super generic mechanism may be overkill, I think typical use cases would be:
* retrieve a single packet (currently supported by this PR)
* discard all packets
* write all packets to a file in the background
* print all packets in a loop until I hit Ctrl-C: this can be achieved with the current implementation from @Yi-Tseng so nothing is needed here besides an example in `usage/` maybe

Originally I thought we would be able to register custom handlers which would apply to all packet-ins or packet-ins matching a specific filter (based on metadata fields), but that may be out-of-scope for a shell. Although the ability to redirect to different files (3rd use case above) based on the value of certain metadata fields could be neat.

Commit:
address comments

Yi-Tseng commented:
Hi @antoninbas, thanks for the review
I totally agree that we can add support to register some call-back functions/handlers to handle packet-ins. I should be able to include it in this PR.

antoninbas commented on README.md:
```diff
@@ -93,6 +93,8 @@ about P4Info objects.
 P4Runtime `Entity` fields), along with `multicast_group_entry` and
 `clone_session_entry`, for runtime entity programming.
 
+`packet` is the command for packet IO, see the [usage](usage/packet_io.md) for more information.
```
needs to be updated now that we have different commands for `packet_in` / `packet_out`

antoninbas commented on p4runtime_sh/context.py:
```diff
@@ -55,6 +56,11 @@ class P4RuntimeEntity(enum.Enum):
     packet_replication_engine_entry = 8
 
 
+@enum.unique
+class P4RuntimeStreamMessage(enum.Enum):
+    packet = 1
```
you don't seem to be using this anymore?

antoninbas commented on p4runtime_sh/shell.py:
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
+        for name, info in self._md_info.items():
+            self.__doc__ += str(info)
+        self.__doc__ += """
+Set a metadata value with <self>['<metadata_name>'] = '...'
```
```suggestion
Set a metadata value with <self>.['<metadata_name>'] = '...'
```

antoninbas commented on p4runtime_sh/shell.py:
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
+latest_hid = 0
+packet_in_handles = {}  # Handle ID -> handle function
+class PacketIn():
+
```
extra empty whiteline?

antoninbas commented on p4runtime_sh/shell.py:
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
+        return self._md.values()
+
+
+latest_hid = 0
+packet_in_handles = {}  # Handle ID -> handle function
```
could we wrap this in a class to and have a single global object to manage handlers?

antoninbas commented on usage/packet_io.md:
```diff
@@ -0,0 +1,140 @@
+### Set up handlers to handle different types of packet-in
+
+To handle the packet-in asynchronously, we can create handlers/functions to process
+packet-ins once we received them.
```
s/once we received them./as we receive them.

antoninbas commented on usage/packet_io.md:
````diff
@@ -0,0 +1,140 @@
+packet-ins once we received them.
+
+```python
+# Create a handle to print every packet-in messages
+P4Runtime sh >>> def handle(x):
+            ...:     print(x)
+P4Runtime sh >>> packet_in.register_handler(handle)
````
s/handle/handler ?

antoninbas commented on p4runtime_sh/shell.py:
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
+        latest_hid += 1
+        hid = latest_hid
+        packet_in_handles[hid] = _handle_wraper
+        return hid
```
nit: in addition to returning the handler id, maybe we could add an attribute to the handler function object (`handler.hid = <hid>`) to make unregistering the handler easier

antoninbas commented on p4runtime_sh/shell.py:
```diff
@@ -2263,6 +2280,195 @@ def add(self, egress_port=None, instance=0):
+        self.recv_t = Thread(target=_packet_in_recv_func, args=(self, ))
+        self.recv_t.start()
+
+    def register_handler(self, handle, payload_regex=b'', **kwargs):
+        # Register a packet-in handle, will return the handle ID.
+        # Use keyword argument as additional filter to filter the packet in
```
why "handle" instead of "handler"?

antoninbas commented on usage/packet_io.md:
```diff
@@ -0,0 +1,140 @@
+
+The `sniff` function will return a list which contains packet-in messages when:
+
+- Receive enough number packets(by passing the `count` parameter)
+- Expired(by passing the `timeout` parameter)
+- Keyboard interrupt(Ctrl + C)
```
```suggestion
- Enough packets have been received (based on the `count` parameter)
- The timeout expires (based on the `timeout` parameter)
- A keyboard interrupt occurs (Ctrl + C)
```

antoninbas commented on usage/packet_io.md:
````diff
@@ -0,0 +1,140 @@
+- Keyboard interrupt(Ctrl + C)
+
+```python
+P4Runtime sh >>> packet_in.sniff(timeout=1, count=1)
````
maybe this should return an iterator?

antoninbas commented on usage/packet_io.md:
```diff
@@ -0,0 +1,140 @@
+ }]
+
+# Both timeout and count can be "None".
+# It will wait until user send a keyboard interrupt(Ctrl + C) to the shell.
```
s/send/sends

antoninbas commented on usage/packet_io.md:
```diff
@@ -0,0 +1,140 @@
+x = packet_in.sniff(to_file='/tmp/pkt-in.txt')
+
+# To show packet-in messages, set `verbose` to `True`
+x = packet_in.sniff(verbose=True)
```
not sure `verbose` is necessary. If we use an iterator instead, user can write:
```
for x in packet_in.sniff():
    print(x)
```
making it similar to how support read operations for entities

antoninbas commented on usage/packet_io.md:
```diff
@@ -0,0 +1,140 @@
+# matches "AA.*A" in payload with ingress port 0, for example "AAbA".
+P4Runtime sh >>> def handle(x):
+            ...:     print(x)
+P4Runtime sh >>> packet_in.register_handler(handle, payload_regex=b'AA.*A', ingress_port='0')
```
I believe the idea was to support filtering based on controller metadata fields primarily

Commit:
Address comments and cleanup

 - `sniff` function now return an iterator of packet-in message instead of
   a list.
 - Remove count from `sniff` function
 - Add `function` parameter to `sniff` function and remove handler
   mechanism from the PacketIn class
 - Add unit test for PacketOut

Commit:
Merge remote-tracking branch 'upstream/main' into stream-msg

Yi-Tseng commented:
Sorry for the delay, 
I resolved some comments and simplify the PacketIn class, the sniff function is more like the read function from other class (returns an iterator by default, and can also pass a function)

antoninbas commented on p4runtime_sh/shell.py:
```diff
@@ -616,6 +633,7 @@ def __init__(self, entity_type, p4runtime_cls, modify_only=False):
         self._entity_type = entity_type
         self._entry = p4runtime_cls()
         self._modify_only = modify_only
+        self._p4runtime_cls = p4runtime_cls
```
Is `self._p4runtime_cls` used anywhere?

antoninbas commented on p4runtime_sh/test.py:
```diff
@@ -1109,6 +1110,48 @@ def test_global_options_invalid(self):
+        packet_out.metadata['egress_port'] = '1'
+        packet_out.send()
+
+        actual_msg = sh.client.stream_out_q.get()
```
how can this work consistently? I would expect the thread executing `stream_req_iterator` in `P4RuntimeClient` to potentially remove the item from the queue (and send it to the server) before we can retrieve it here. Isn't there a potential race condition?

Yi-Tseng replied:
Right, I think we should store the packet-out message in the mock server(Servicer) and get the message from there.

Commit:
Remove unused parts and rewrite packet-out test

antoninbas commented on p4runtime_sh/test.py:
```diff
@@ -67,6 +70,9 @@ def StreamChannel(self, request_iterator, context):
                 yield rep
+            elif req.HasField('packet'):
+                self.stored_packet_out.put(req)
+                yield p4runtime_pb2.StreamMessageResponse()
```
I don't think this yield statement is needed / makes sense. When a P4Runtime server receives a PacketOut message, it doesn't need to reply with an empty stream message response.

Yi-Tseng replied:
Right, I think I put this here by mistake

antoninbas reviewed (commented):
one more nit, otherwise LGTM

Commit:
address comment
